src/rsync.rs

```diff
@@ -1,76 +1,256 @@
+fn symlink_current_to_new_revision_dir(
+    new_revision: &RsyncRevision,
+    config: &Config,
+) -> Result<()> {
     info!(
-        "Writing Rsync repository to: {}",
-        out_path.to_string_lossy()
+        "Updating symlink 'current' to '{}' under rsync dir '{:?}'",
+        new_revision.dir_name(),
+        config.rsync_dir
     );
-    write_rsync_content(&out_path, notify, client, raw_snapshot)?;
-
-    if cfg!(unix) {
-        info!("Use symlink to link rsync module dir to the new content");
-        // create a new symlink then rename it
-        let tmp_name = file_ops::set_path_ext(&opt.rsync_dir, config::TMP_FILE_EXT);
-        std::os::unix::fs::symlink(&out_path, &tmp_name)?;
-        std::fs::rename(&tmp_name, &opt.rsync_dir)?;
-    } else {
-        info!("Renaming rsync folders for close to atomic update of the rsync module dir");
-        file_ops::install_new_dir(&opt.rsync_dir, last_serial.to_string())?;
+    let current_path = config.rsync_dir_current();
+
+    let tmp_name = file_ops::path_with_extension(&current_path, config::TMP_FILE_EXT);
+    if tmp_name.exists() {
+        std::fs::remove_file(&tmp_name).with_context(|| {
+            format!(
+                "Could not remove lingering temporary symlink for current rsync dir at '{:?}'",
+                tmp_name
+            )
+        })?;
     }
 
+    std::os::unix::fs::symlink(new_revision.dir_name(), &tmp_name).with_context(|| {
+        format!(
+            "Could not create temporary symlink for new rsync content at '{:?}'",
+            tmp_name
+        )
+    })?;
+
+    std::fs::rename(&tmp_name, &current_path).with_context(|| {
+        format!(
+            "Could not rename symlink for current rsync dir from '{:?}' to '{:?}'",
+            tmp_name, current_path
+        )
+    })?;
+
     Ok(())
 }
 
-fn write_rsync_content(
-    out_path: &Path,
-    notify: &mut NotificationFile,
-    client: &HttpClient,
-    raw_snapshot: &[u8],
+/// Rename the path for the new revision to the current rsync path, *after*
+/// renaming any existing current path to the serial and session for that
+/// revision.
+fn rename_new_revision_dir_to_current(
+    new_revision: &RsyncRevision,
+    rsync_state: &RsyncDirState,
+    config: &Config,
 ) -> Result<()> {
-    client
-        .snapshot_from_buf(
-            &notify,
-            |uri| {
-                let this_out_path = out_path.join(make_rsync_repo_path(&uri).unwrap());
-                trace!("Writing Rsync file {:?}", &this_out_path);
-                this_out_path
-            },
-            &raw_snapshot,
+    info!("Renaming rsync folders for close to atomic update of the rsync module dir");
+
+    let current_path = config.rsync_dir_current();
+
+    if let Some(current) = &rsync_state.current {
+        let current_preserve_path = current.path(config);
+
+        if current_path.exists() {
+            info!(
+                "Backing up rsync directory for previous revision to: {:?}",
```

Comment:
I misunderstood this. To me "backing up" means making a copy of. Here we do not make a copy, rather we keep the existing copy but under a new name and rename a new copy to the old name. Also the error doesn't say it couldn't backup but that it couldn't rename. I suggest to refer to renaming rather than "backing up".

Reply:
indeed, fixed